### ISSUE TITLE
Configurable Header Insertions

### ISSUE THREAD

**Issue by njames93** [clangd/clangd#1367]
### 1
I'm working on a project where the style dictates that all the includes from our public SDK are included using angle brackets.
However when we build the SDK the headers are included with `-I` instead of `-isystem`.
This results in clangd trying to use quoted includes instead of angle brackets.
```cpp
#include "sdk/example.h"

#include <sdk/example.h>
```

### 2
This is more of a build configuration issue, but I have noticed sometimes when a library is installed in multiple places, both of which are accessible from the system include path, clangd will have to choose and often it may not choose the library version you were after.
```cpp
#include <library-10/example.h>
#include <library/example.h>
```

It would be nice if there was a way to configue this so we could specify how to include these headers.

## Design criteria

Define a new enum in the config called `IncludeType` containing:
 - `Quoted`
 - `Angled`
 - `Auto`
 
Define a new type in the config called `IncludeRule`, containing:
 - `llvm::Regex PathMatch`
 - `IncludeType Type = Auto`
 - `llvm::Optional<std::string> Replace`
 
Add a new block to the config called `IncludeRules` which would be a vector of `IncludeRule`.

When it comes to creating include fixes, we check each IncludeRule and when we find a first match:
 If IncludeType is not `Auto`, update the include type that clangd wants to emit.
If Replace is valid, perform a regex replacement of the include.

The specifics or `PathMatch` aren't set it stone, it may be preferable to require complete matching of the path, but I personally think partial matching should be more than enough 

As this vector is traversed in order until the first match there should be a sensible ordering based on the source of the config.
Ideally Rules in a project root should have lower priority than rules in the directory where the file being worked on lives. For multiple fragments in, I'd argue it makes most sense if the rules appear in fragment order.

For example with the config 
```yaml
IncludeRules:
  - PathMatch: '^library-10/'
  - Type: Quoted
  - Replace: 'library/'
```
For include `<library-10/example.h>` clangd would instead suggest `"library/example.h"`

**Comment by HighCommander4** [comment]
See also #1247 for configuring quoted vs. angled in particular.

**Comment by kadircet** [comment]
I can see the confusion inside clangd around quoted vs angled brackets, and that's something that has been brought up before. But hasn't been high priority and we've been advising people for indicating their intent explicitly inside the compile flags instead (as you've also pointed out).

Surely we can perform a remapping as you've proposed here, but it's important to not forget that include-insertion calculations happen eagerly today for every code completion candidate. So performing regex searches and rewrites there might not be feasible at request serving path.
Another alternative would be to infer that information directly during indexing and incorporate into the index as a signal (when inserting prefer quoted/system).

I am not following the issue around multiple versions of the same library though (and I don't think a rewrite during code-completion triggered include insertion would solve it for rest of the functionality, e.g. go-to-def will likely still take you to wrong place, xrefs would be misleading).
Do you have multiple versions of a library accessed by the same project, hence clangd's index contains symbols for both versions or do you have a symlink somewhere pointing to a versioned directory in your compile flags and clangd is resolving the symlink and still pointing at the "right" directory but with the wrong spelling?
For such a scenario I think it's reasonable to try and figure out a solution (possibly based on symlink resolution, rather than adding more configuration) but I am not sure if multiple versions of same library existing inside a single project is feasible to solve as things stand today.

**Comment by njames93** [comment]
> I can see the confusion inside clangd around quoted vs angled brackets, and that's something that has been brought up before. But hasn't been high priority and we've been advising people for indicating their intent explicitly inside the compile flags instead (as you've also pointed out).

That is a good intermediate step, however for complicated projects This would require changing a lot of compile flags and if those directories are absolute, it would be even more of a pain.
The project I'm working on has up to 30 library includes per each subproject. As there is no way to dynamically convert `-I` to `-isystem` based on the path it would require hand crafting a configuration file for each project, then updating the configuration file should an issue occur.
> 
> Surely we can perform a remapping as you've proposed here, but it's important to not forget that include-insertion calculations happen eagerly today for every code completion candidate. So performing regex searches and rewrites there might not be feasible at request serving path. Another alternative would be to infer that information directly during indexing and incorporate into the index as a signal (when inserting prefer quoted/system).

It's my understanding that the include insertion for code complete only happens when the server tries to complete, os for most of the time this wouldn't even run.
We could also potentially cache results if you see this being a large performance concern.
Storing the information in the index has promise, but lacks the granularity of different subprojects wanting slightly different include mapping logic.

> 
> I am not following the issue around multiple versions of the same library though (and I don't think a rewrite during code-completion triggered include insertion would solve it for rest of the functionality, e.g. go-to-def will likely still take you to wrong place, xrefs would be misleading). Do you have multiple versions of a library accessed by the same project, hence clangd's index contains symbols for both versions or do you have a symlink somewhere pointing to a versioned directory in your compile flags and clangd is resolving the symlink and still pointing at the "right" directory but with the wrong spelling? For such a scenario I think it's reasonable to try and figure out a solution (possibly based on symlink resolution, rather than adding more configuration) but I am not sure if multiple versions of same library existing inside a single project is feasible to solve as things stand today.

Probably not the best example to use for the replacing logic. The issue I had is a system library on my system include path was getting used instead. However the more common is when clangd tries to insert a private implementation header file instead of the public one that exports is. I know clangd does have some IWYU logic but if a system header doesn't use any kind of hints it can get rather annoying having to correct it each time, I know people who have just turned off the header includes outright because of these issues.

**Comment by kadircet** [comment]
> As there is no way to dynamically convert -I to -isystem based on the path it would require hand crafting a configuration file for each project

Right. The idea there is definitely not about customizing these with clangd config fragments, but rather making it more clear at the build system level so that compile flags provided catch this intent better.

> It's my understanding that the include insertion for code complete only happens when the server tries to complete, os for most of the time this wouldn't even run.

not sure what you mean by server tries to complete. i guess you meant user, if so, yes include insertion only happens when the user accepts a completion. but clangd, in its current infra, pre-computes all the edits in advance for those header insertions.

> We could also potentially cache results if you see this being a large performance concern.

i don't really see how we can do this as in the end we need to run a regex match for each completion suggestion with a header (re-writing already happens, it won't be much of an extra cost).  but maybe i am exaggerating and this extra regex match per completion item isn't such a big deal in practice (considering all the other operations we do per header).

> Storing the information in the index has promise, but lacks the granularity of different subprojects wanting slightly different include mapping logic.

right, that's definitely a bigger concern if there are different conventions per subdirectory, despite all of the sources being part of the same compilation database.

> The issue I had is a system library on my system include path was getting used instead.

Wouldn't your builds also use that include path then?

>  However the more common is when clangd tries to insert a private implementation header file instead of the public one that exports is. I know clangd does have some IWYU logic but if a system header doesn't use any kind of hints it can get rather annoying having to correct it each time, I know people who have just turned off the header includes outright because of these issues.

Right, this is a bigger issue that we are actively working on nowadays. There are 2 ways to address this issue today (not fully implemented though, so more like concepts). you can either have `IWYU private` pragmas in the private header and point at the public one (as you pointed out this is not an option in cases where you don't own the header). The latter is putting `IWYU export` pragmas to the public header (but this won't really tell clangd to consider that public header as the main provider).
So rather than re-writing, maybe the right solution would be to just say don't insert headers that match a regex or under a certain directory. Because re-writing include-paths seem like a lot more complicated and would probably require a bigger design (especially analysis on what kind of problems it's trying to solve, as this doesn't seem to be a big enough concern to warrant such a mechanism today).

**Comment by sam-mccall** [comment]
As I just mentioned on the code review, I think this design is very complicated for the value it provides. I don't see evidence that there's widespread need for a general include rewrite system.

For `<>` vs `""` in particular, we do have multiple reports suggesting this. I think we should be open to a either a **simple** config-based solution, or one that infers the quoting style at indexing time (though I think the latter is both complex to implement and risky).

I think we should close this in favor of #1247, which is much more narrowly and appropriately scoped.

It would be useful to have some more info about the use case here though...

> However when we build the SDK the headers are included with -I instead of -isystem.
> ...
> This would require changing a lot of compile flags and if those directories are absolute, it would be even more of a pain.

Can you explain concretely why changing `-I` to `-isystem` for headers you want to include with `<brackets>` is hard or problematic, and what absolute paths have to do with it? I think it's clear that specifying this hint at the origin point is *conceptually* simpler than trying to match and adjust them later.

**Comment by rhd** [comment]
This change would probably help users that are using [spack environments](https://spack.readthedocs.io/en/latest/environments.html). 

[Spack](https://spack.io/) "is a package manager for supercomputers, Linux, and macOS."

The current issue is that when you use a "spack environment", clangd injects a pattern like `#include ".spack-env/view/<hash>/spdlog/spdlog.h"` instead of simply `#include <spdlog/spdlog.h>`.

It seems these rewrite rules would help solve this specific problem as well. With that said, a simple "ignore this directory" would also probably work.

**Comment by dmpolukhin** [comment]
We have a similar issue with paths that clangd uses when inserting headers during autocompletion. They don’t follow our style guide and in general do not always work because inserting a header may not be enough; the dependency has to be added to the build system. Another example of what might be wrong when clangd uses symlink from intermediate build directory instead of canonical path to the source in the repo. In general the logic which path should be used is very specific to our environment so implementing in the upstream doesn’t make sense. Proposed solution with text based replacement is somewhat limited and may not cover all our use cases so we are thinking about adding a config option that will specify a custom script that clangd could call to transform full absolute path to the header that clangd want to insert into the form that matches user expectations about header spelling.

**Comment by HighCommander4** [comment]
@dmpolukhin Do you have any ideas about the performance concern discussed in earlier comments?

To elaborate on it a bit: header insertion is implemented by placing a text edit describing the header insertion into the `additionalTextEdits` field of the LSP [`CompletionItem`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem). This means those text edits (including the spelling of the include path that will go in them) need to be computed at the time completion is requested (as opposed to at the time a completion proposal is selected by the user), and they need to be computed for every completion item.

So, if you're invoking completion at the `^` in `#include <foo/^`, and the `foo` directory contains 100 header files, with the current architecture clangd would need to make 100 invocations to the script as part of building the response to the `textDocument/completion` request, which I think is unrealistic.

**Comment by dmpolukhin** [comment]
@HighCommander4 It seems that the main problem with performance happens due to not using `completionItem/resolve` requests and calculating all of them eagerly. So it seems that it is the pre-requisite for any non-trivial include path processing. Thank you for highlighting the performance issue, it makes the whole task bigger than I thought.

**Comment by HighCommander4** [comment]
> the main problem with performance happens due to not using `completionItem/resolve` requests

I looked into this a bit, and it seems like an interesting approach, though I can think of two potential issues with using `completionItem/resolve` for this purpose:

 1. It seems to be geared towards computing fields like `details` and `documentation`, and so the editor calls it whenever a completion item is focused in the completion popup (e.g. when you press the down arrow key to view the details of the next completion item). That still seems a bit more eager than what would be ideal, which would be to invoke the script only once the completion item is actually chosen for insertion.
 2. It looks like editors need to opt into which properties of `CompletionItem` can be computed by `resolve`. VSCode does support `additionalTextEdits`, but I haven't checked other editors.

I did notice another field of `CompletionItem` that seems like it could be a better fit:

```
       /**
	 * An optional command that is executed *after* inserting this completion.
	 * *Note* that additional modifications to the current document should be
	 * described with the additionalTextEdits-property.
	 */
	command?: Command;
```

in that this will only run once the completion item is inserted.

I do wonder about the note that says "additional modifications to the current document should be described with the additionalTextEdits-property" -- is its intention just to point out that using `additionalTextEdits` is a more efficient way of performing an additional modification to the current document (in that it avoids an extra round-trip to the server to execute the command), or is there some more subtle problem that would come with using `command` to perform an additional text edit?

**Comment by dmpolukhin** [comment]
[Command](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#command) seems like a completely unspecified thing so we can define it with any semantic and commands but it might create future compatibility issues. I was thinking about `command` for implementing modification in other files like build configs but it is vague at the moment and also will be very specific to the environment so calling a script would be the generic implementation.

In general I see a bunch of things that can be customized with making external calls from clangd. So if it's something interesting to have in Clangd upstream, I can write some RFC for the interface. Below are list of customisations on top of my head:
- Fixing spelling of include paths inserted by clangd.
- Missing include paths suggestions when clang handles include files not found. Include file not-found errors usually rune compilation and introduces tons of other false compilation errors that break code navigation. In many cases it is possible to implement some logic to find which files user wants to include so clangd could show errors about missing include but still include the required file to have good language service when possible.
- Custom CDB extrapolation for headers.
- Lazy CDB generation and providing better fall-back compilation flags.

**Comment by HighCommander4** [comment]
> [Command](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#command) seems like a completely unspecified thing so we can define it with any semantic and commands but it might create future compatibility issues. I was thinking about `command` for implementing modification in other files like build configs but it is vague at the moment and also will be very specific to the environment so calling a script would be the generic implementation.

To clarify, I was thinking of a command implemented on the server side.

So, if clangd is configured to use a script to rewrite inserted header spellings, then instead of computing an `additionalTextEdit` for the header insertion, it could give the completion item a `command` with an identifier like `"clangd:insertHeader"` and an argument storing the header path. Then, when the completion item is inserted:

 * the client calls `workspace/executeCommand` with `{command: "clangd:insertHeader", arguments: ["<headerpath>"]}`
 * the server calls the configured script with `<headerpath>` as input
 * the server calls `workspace/applyEdit` to apply an edit to insert the header using the spelling provided by the script

> In general I see a bunch of things that can be customized with making external calls from clangd. So if it's something interesting to have in Clangd upstream, I can write some RFC for the interface.

My personal opinion is that this is potentially interesting, but it's probably wise for wait for feedback from @sam-mccall or @kadircet before investing significant time into a particular direction.

---

Note also that since we're talking about having clangd invoke other executables, security considerations may be in play.

There is one situation in which clangd currently invokes another executable: running a compiler driver to query it for its built-in include paths.

In this situation, the path of the invoked executable needs to be allow-listed using the `--query-driver` command-line option. Based on previous discussions, my understanding is that this is considered a more privileged place to have such an allow-list than clangd's config file. (I guess because in some editors, command line arguments to the language server cannot be provided in a file in the workspace, while clangd config files can be.)

If we're introducing new scenarios where clangd can invoke an executable, we may need to have a similar allow-list in a similarly privileged place.

**Comment by mortenfc** [comment]
Seems like a duplicate of
https://reviews.llvm.org/D145843

Clangd IWYU should just support
https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUMappings.md

**Comment by HighCommander4** [comment]
> Seems like a duplicate of https://reviews.llvm.org/D145843

Not an exact duplicate -- the patch implements a subset of the configuration discussed in this issue (concerning `<>` vs. `""` syntax specifically), as tracked in https://github.com/clangd/clangd/issues/1247.

Note that there is a newer version of that patch at https://github.com/llvm/llvm-project/pull/67749, which I've reviewed and which is waiting on some (relatively minor) updates.

> Clangd IWYU should just support https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUMappings.md

Yes, that would be nice. We have https://github.com/clangd/clangd/issues/1276 on file about that.

**Comment by ChillerDragon** [comment]
This should be at least partially solved by https://github.com/llvm/llvm-project/pull/67749
